File: clients/dotnet/src/Firstrun/Transport.cs
using System;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Firstrun
{
    internal enum SendOutcome
    {
        /// <summary>Accepted. Drop the batch.</summary>
        Accepted,

        /// <summary>The server will say no again in an hour. Drop the batch, do not retry.</summary>
        Rejected,

        /// <summary>Offline, timed out, rate limited, or the server is having a bad day. Keep the batch.</summary>
        Transient,
    }

    internal readonly struct SendResult
    {
        internal SendResult(SendOutcome outcome, string detail, TimeSpan? retryAfter = null)
        {
            Outcome = outcome;
            Detail = detail;
            RetryAfter = retryAfter;
        }

        internal SendOutcome Outcome { get; }
        internal string Detail { get; }

        /// <summary>What the server asked us to wait, when it said so.</summary>
        internal TimeSpan? RetryAfter { get; }
    }

    /// <summary>
    /// One POST to <c>/v1/e</c>.
    /// </summary>
    /// <remarks>
    /// A single <see cref="HttpClient"/> for the life of the client, never one per call:
    /// a per-call HttpClient burns a socket per batch and leaves it in TIME_WAIT, which
    /// is how a telemetry library ends up exhausting the ports of the app it is measuring.
    ///
    /// The response body is read and discarded. There is nothing in it the client can
    /// act on that the status code has not already said, and reading it into a string is
    /// bounded work we do not need to do.
    /// </remarks>
    internal sealed class Transport : IDisposable
    {
        private readonly HttpClient _http;
        private readonly bool _ownsHttp;
        private readonly string _url;

        internal Transport(FirstrunOptions options)
        {
            _url = options.Host.TrimEnd('/') + "/v1/e";

            if (options.HttpClient != null)
            {
                _http = options.HttpClient;
                _ownsHttp = false;
                return;
            }

            _ownsHttp = true;
#if NET8_0_OR_GREATER
            var handler = new SocketsHttpHandler
            {
                ConnectTimeout = options.ConnectTimeout,
                // A telemetry client makes one request every few seconds at most. Holding
                // a pooled connection open for minutes across an idle app is a socket the
                // host is not using for its own work.
                PooledConnectionIdleTimeout = TimeSpan.FromSeconds(90),
                MaxConnectionsPerServer = 2,
                AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate,
            };
            _http = new HttpClient(handler, disposeHandler: true);
#else
            _http = new HttpClient();
#endif
            _http.Timeout = options.RequestTimeout;
        }

        internal async Task<SendResult> SendAsync(string body, CancellationToken cancellation)
        {
            try
            {
                using (var request = new HttpRequestMessage(HttpMethod.Post, _url))
                {
                    request.Content = new StringContent(body, new UTF8Encoding(false), "application/json");
                    // No cookies, no auth, nothing identifying. The source key in the body
                    // is the whole of what we tell the server about the caller.
                    request.Headers.TryAddWithoutValidation("User-Agent", "firstrun-dotnet");

                    using (HttpResponseMessage response = await _http
                        .SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellation)
                        .ConfigureAwait(false))
                    {
                        int code = (int)response.StatusCode;

                        if (code >= 200 && code < 300)
                        {
                            return new SendResult(SendOutcome.Accepted, code.ToString());
                        }

                        // 408 and 429 are the two 4xx that mean "later", not "never".
                        if (code == 408 || code == 429)
                        {
                            return new SendResult(SendOutcome.Transient, "http " + code, RetryAfter(response));
                        }

                        if (code >= 400 && code < 500)
                        {
                            // A malformed batch, or a source key that no longer exists.
                            // Retrying forever would wedge every later event behind it.
                            return new SendResult(SendOutcome.Rejected, "http " + code);
                        }

                        return new SendResult(SendOutcome.Transient, "http " + code, RetryAfter(response));
                    }
                }
            }
            catch (OperationCanceledException) when (cancellation.IsCancellationRequested)
            {
                return new SendResult(SendOutcome.Transient, "cancelled");
            }
            catch (OperationCanceledException)
            {
                // HttpClient reports its own timeout as a cancellation on most runtimes.
                return new SendResult(SendOutcome.Transient, "timeout");
            }
            catch (HttpRequestException ex)
            {
                return new SendResult(SendOutcome.Transient, Describe(ex));
            }
            catch (Exception ex)
            {
                // Whatever this was, it is not worth an unhandled exception on a
                // background thread of somebody else's application.
                return new SendResult(SendOutcome.Transient, Describe(ex));
            }
        }

        private static TimeSpan? RetryAfter(HttpResponseMessage response)
        {
            try
            {
                System.Net.Http.Headers.RetryConditionHeaderValue? header = response.Headers.RetryAfter;
                if (header == null) return null;
                if (header.Delta.HasValue) return header.Delta.Value;
                if (header.Date.HasValue)
                {
                    TimeSpan delta = header.Date.Value - DateTimeOffset.UtcNow;
                    return delta > TimeSpan.Zero ? delta : TimeSpan.Zero;
                }
            }
            catch
            {
                // A malformed header is not worth a failed send path.
            }
            return null;
        }

        private static string Describe(Exception ex)
        {
            Exception root = ex;
            while (root.InnerException != null) root = root.InnerException;
            return root.GetType().Name + ": " + root.Message;
        }

        public void Dispose()
        {
            if (_ownsHttp) _http.Dispose();
        }
    }
}

File: clients/python/src/firstrun/_transport.py
"""One POST to ``/v1/e``, using only the standard library.

``urllib.request`` rather than ``requests`` because a telemetry library has no
business adding a dependency to the program it is measuring, and no business
being the reason two packages disagree about a urllib3 version.

The cost is that urllib opens a fresh connection per batch rather than pooling
one. At one batch every few seconds that is a handshake we can afford, and it
removes the whole class of bugs where a pooled socket went stale during an
outage and the first send after recovery fails for no visible reason.
"""

from __future__ import annotations

import json
import ssl
import urllib.error
import urllib.request
from typing import Any, Dict, List, NamedTuple, Optional

ACCEPTED = "accepted"
"""The server took it. Drop the batch."""

REJECTED = "rejected"
"""The server will say no again in an hour. Drop the batch, do not retry."""

TRANSIENT = "transient"
"""Offline, timed out, rate limited, or a bad day at the server. Keep the batch."""


class SendResult(NamedTuple):
    outcome: str
    detail: str
    retry_after: Optional[float] = None


def build_batch(
    source_key: str,
    resource: Optional[Dict[str, Any]],
    entries: List[Dict[str, Any]],
) -> bytes:
    """The LogBatch body, as UTF-8 JSON.

    The keys are one letter because this is the same body the browser tag posts
    from ``sendBeacon`` on a page being unloaded, where bytes are the constraint:
    one shape for every client rather than a compact browser dialect beside a
    verbose SDK one. ``k`` is the source key, ``r`` the resource and ``e`` the
    entries. There is no top-level id field: identity is three optional
    attributes and they travel in ``r`` like everything else about the client.

    ``r`` carries what is true of the whole PROCESS rather than of one entry: the
    service, the build, the operating system. It sits once per body because it
    does not change between two entries in the same request, and repeating it 200
    times is 200 copies of the same strings. The edge merges it UNDER each
    entry's own attributes, so an entry that sets the same key wins.

    Keys the server does not define are never added, and an empty resource is
    left out rather than sent as an empty object.

    Source of truth: ``LogBatch`` in ``packages/schema/src/log.ts``.
    """
    body: Dict[str, Any] = {"k": source_key, "e": entries}
    if resource:
        body["r"] = resource

    # allow_nan=False so a NaN that slipped through the attribute cleaner raises
    # here, where the caller can drop one batch, rather than being written as the
    # bare literal `NaN` that the server's JSON parser rejects along with every
    # entry travelling beside it.
    return json.dumps(
        body, separators=(",", ":"), ensure_ascii=False, allow_nan=False
    ).encode("utf-8")


class Transport:
    """Sends one batch. Returns a result; never raises."""

    def __init__(self, host: str, timeout: float, ssl_context: Optional[ssl.SSLContext] = None) -> None:
        self.url = host.rstrip("/") + "/v1/e"
        self.timeout = timeout
        self._opener = urllib.request.build_opener(
            urllib.request.HTTPSHandler(context=ssl_context) if ssl_context else urllib.request.HTTPSHandler(),
            # No redirects. An analytics POST that gets 302'd somewhere is a
            # misconfiguration, not something to follow.
            _NoRedirect(),
        )

    def send(self, body: bytes) -> SendResult:
        request = urllib.request.Request(
            self.url,
            data=body,
            method="POST",
            headers={
                "Content-Type": "application/json",
                "Content-Length": str(len(body)),
                # No cookies, no auth, no identifying header. The source key in
                # the body is the whole of what we tell the server about us.
                "User-Agent": "firstrun-python",
                "Connection": "close",
            },
        )

        try:
            with self._opener.open(request, timeout=self.timeout) as response:
                # Read and discard. There is nothing in the body the client can
                # act on that the status code has not already said.
                try:
                    response.read(4096)
                except Exception:
                    pass
                status = int(getattr(response, "status", None) or response.getcode() or 0)
                if 200 <= status < 300:
                    return SendResult(ACCEPTED, str(status))
                return SendResult(TRANSIENT, "http %d" % status)

        except urllib.error.HTTPError as exc:
            status = int(exc.code)
            retry_after = _retry_after(exc)
            try:
                exc.read()
                exc.close()
            except Exception:
                pass

            # A redirect we refused to follow. The host is misconfigured, and it
            # will still be misconfigured in an hour, so this is not a retry.
            if 300 <= status < 400:
                return SendResult(REJECTED, "http %d (redirect)" % status)

            # 408 and 429 are the two 4xx that mean "later", not "never".
            if status in (408, 429):
                return SendResult(TRANSIENT, "http %d" % status, retry_after)
            if 400 <= status < 500:
                # A malformed batch, or a source key that no longer exists.
                # Retrying it forever would wedge every later event behind it.
                return SendResult(REJECTED, "http %d" % status)
            return SendResult(TRANSIENT, "http %d" % status, retry_after)

        except Exception as exc:
            # Offline, DNS failure, TLS failure, timeout, a proxy having
            # opinions. All of it is "try again later" and none of it is the
            # host program's problem.
            return SendResult(TRANSIENT, "%s: %s" % (type(exc).__name__, exc))


def _retry_after(exc: "urllib.error.HTTPError") -> Optional[float]:
    try:
        raw = exc.headers.get("Retry-After") if exc.headers else None
        if not raw:
            return None
        return float(int(str(raw).strip()))
    except Exception:
        # A date-form Retry-After, or nonsense. The backoff covers us either way.
        return None


class _NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):  # type: ignore[no-untyped-def]
        return None
